### PR TITLE
Get all addresses when resolving hostname

### PR DESCRIPTION
**Pull request by gotnospirit**
For a test I had to resolve a hostname and randomly pick one address but IP::resolve_hostname only returns the first address.
I've seen something were done for issue #414 but it seems like the code wasn't merged.
So I've implement two new methods :
`Array IP::resolve_hostname_addresses(const String &p_hostname, Type p_type)`
`Array IP::get_resolve_addresses(ResolverID p_id) const`

Any comments are welcome.

### CODE REVIEW

File: drivers/unix/ip_unix.h [hunk 1]
```diff
@@ -37,13 +37,11 @@
 class IP_Unix : public IP {
 	OBJ_TYPE(IP_Unix, IP);
 
-	virtual IP_Address _resolve_hostname(const String &p_hostname, IP::Type p_type);
-
 	static IP *_create_unix();
-
+
```
Comment (akien-mga) [inline drivers/unix/ip_unix.h]: Trailing whitespace.

**Comment by akien-mga** [comment]
CC @Faless

**Review by Faless — CHANGES_REQUESTED** [review]
So @gotnospirit ... it took me 2 months but I finally found the time to review this. I'm sorry for the long delay, I know having a PR pending for months is frustrating. Again, sorry.

Now, back to business...

For naming consistency we should have `get_resolve_addresses` renamed to `get_resolve_item_addresses`.
See this patch to apply on top of your commits:
```
diff --git a/core/io/ip.cpp b/core/io/ip.cpp
index 09a7795c9..c078a6edc 100644
--- a/core/io/ip.cpp
+++ b/core/io/ip.cpp
@@ -223,7 +223,7 @@ IP_Address IP::get_resolve_item_address(ResolverID p_id) const {
 	return IP_Address();
 }
 
-Array IP::get_resolve_addresses(ResolverID p_id) const {
+Array IP::get_resolve_item_addresses(ResolverID p_id) const {
 
 	ERR_FAIL_INDEX_V(p_id, IP::RESOLVER_MAX_QUERIES, Array());
 
@@ -294,7 +294,7 @@ void IP::_bind_methods() {
 	ObjectTypeDB::bind_method(_MD("resolve_hostname_queue_item", "host", "ip_type"), &IP::resolve_hostname_queue_item, DEFVAL(IP::TYPE_ANY));
 	ObjectTypeDB::bind_method(_MD("get_resolve_item_status", "id"), &IP::get_resolve_item_status);
 	ObjectTypeDB::bind_method(_MD("get_resolve_item_address", "id"), &IP::get_resolve_item_address);
-	ObjectTypeDB::bind_method(_MD("get_resolve_addresses", "id"), &IP::get_resolve_addresses);
+	ObjectTypeDB::bind_method(_MD("get_resolve_item_addresses", "id"), &IP::get_resolve_item_addresses);
 	ObjectTypeDB::bind_method(_MD("erase_resolve_item", "id"), &IP::erase_resolve_item);
 	ObjectTypeDB::bind_method(_MD("get_local_addresses"), &IP::_get_local_addresses);
 	ObjectTypeDB::bind_method(_MD("clear_cache"), &IP::clear_cache, DEFVAL(""));
diff --git a/core/io/ip.h b/core/io/ip.h
index 2d05712c1..1ceaa7583 100644
--- a/core/io/ip.h
+++ b/core/io/ip.h
@@ -81,7 +81,7 @@ public:
 	ResolverID resolve_hostname_queue_item(const String &p_hostname, Type p_type = TYPE_ANY);
 	ResolverStatus get_resolve_item_status(ResolverID p_id) const;
 	IP_Address get_resolve_item_address(ResolverID p_id) const;
-	Array get_resolve_addresses(ResolverID p_id) const;
+	Array get_resolve_item_addresses(ResolverID p_id) const;
 	virtual void _resolve_hostname(List<IP_Address> & r_addresses, const String &p_hostname, Type p_type = TYPE_ANY) const = 0;
 	virtual void get_local_addresses(List<IP_Address> *r_addresses) const = 0;
 	void erase_resolve_item(ResolverID p_id);
diff --git a/doc/base/classes.xml b/doc/base/classes.xml
index f3f57377d..7ab76faaa 100644
--- a/doc/base/classes.xml
+++ b/doc/base/classes.xml
@@ -15782,7 +15782,7 @@
 				Return a resolved item address, or an empty string if an error happened or resolution didn't happen yet (see [method get_resolve_item_status]).
 			</description>
 		</method>
-		<method name="get_resolve_addresses" qualifiers="const">
+		<method name="get_resolve_item_addresses" qualifiers="const">
 			<return type="Array">
 			</return>
 			<argument index="0" name="id" type="int">
```

I noticed that on my machine some IPs are returned twice (or three times), I think only returning unique IP would be a better option.
Can apply this patch to your commits? I think it would make the result more intuitive..
```
diff --git a/drivers/unix/ip_unix.cpp b/drivers/unix/ip_unix.cpp
index 941b07e0c..401b98a8b 100644
--- a/drivers/unix/ip_unix.cpp
+++ b/drivers/unix/ip_unix.cpp
@@ -116,7 +116,13 @@ void IP_Unix::_resolve_hostname(List<IP_Address> & r_addresses, const String &p_
 	struct addrinfo *next = result;
 
 	do {
-		r_addresses.push_back(_sockaddr2ip(next->ai_addr));
+		if (next->ai_addr == NULL) {
+			next = next->ai_next;
+			continue;
+		}
+		IP_Address ip = _sockaddr2ip(next->ai_addr);
+		if (!r_addresses.find(ip))
+			r_addresses.push_back(_sockaddr2ip(next->ai_addr));
 		next = next->ai_next;
 	} while (next);
```

As a final but important note. Style matters, for future patches you should add the `clang-format` hook to git. It automatically do formatting for you when committing, it's a real life saver. This is the patch it generated to format your code:
```
diff --git a/core/io/ip.cpp b/core/io/ip.cpp
index c078a6edc..5221b7479 100644
--- a/core/io/ip.cpp
+++ b/core/io/ip.cpp
@@ -82,8 +82,7 @@ struct _IP_ResolverPrivate {
 			if (queue[i].status != IP::RESOLVER_STATUS_WAITING)
 				continue;
 			IP::get_singleton()->_resolve_hostname(queue[i].response, queue[i].hostname, queue[i].type);
-			queue[i].status = queue[i].response.empty()
-				? IP::RESOLVER_STATUS_ERROR : IP::RESOLVER_STATUS_DONE;
+			queue[i].status = queue[i].response.empty() ? IP::RESOLVER_STATUS_ERROR : IP::RESOLVER_STATUS_DONE;
 		}
 	}
 
@@ -121,7 +120,7 @@ IP_Address IP::resolve_hostname(const String &p_hostname, IP::Type p_type) {
 		_resolve_hostname(res, p_hostname, p_type);
 		resolver->cache[key] = res;
 	}
-	resolver->mutex->unlock();		
+	resolver->mutex->unlock();
 
 	for (int i = 0; i < res.size(); ++i) {
 		if (res[i].is_valid()) {
@@ -140,7 +139,7 @@ Array IP::resolve_hostname_addresses(const String &p_hostname, Type p_type) {
 		_resolve_hostname(resolver->cache[key], p_hostname, p_type);
 	}
 
-	List<IP_Address> res = resolver->cache[key];	
+	List<IP_Address> res = resolver->cache[key];
 	resolver->mutex->unlock();
 
 	Array result;
diff --git a/core/io/ip.h b/core/io/ip.h
index 1ceaa7583..6f4e19a68 100644
--- a/core/io/ip.h
+++ b/core/io/ip.h
@@ -82,7 +82,7 @@ public:
 	ResolverStatus get_resolve_item_status(ResolverID p_id) const;
 	IP_Address get_resolve_item_address(ResolverID p_id) const;
 	Array get_resolve_item_addresses(ResolverID p_id) const;
-	virtual void _resolve_hostname(List<IP_Address> & r_addresses, const String &p_hostname, Type p_type = TYPE_ANY) const = 0;
+	virtual void _resolve_hostname(List<IP_Address> &r_addresses, const String &p_hostname, Type p_type = TYPE_ANY) const = 0;
 	virtual void get_local_addresses(List<IP_Address> *r_addresses) const = 0;
 	void erase_resolve_item(ResolverID p_id);
 
diff --git a/drivers/unix/ip_unix.cpp b/drivers/unix/ip_unix.cpp
index 401b98a8b..2377fa898 100644
--- a/drivers/unix/ip_unix.cpp
+++ b/drivers/unix/ip_unix.cpp
@@ -85,7 +85,7 @@ static IP_Address _sockaddr2ip(struct sockaddr *p_addr) {
 	return ip;
 };
 
-void IP_Unix::_resolve_hostname(List<IP_Address> & r_addresses, const String &p_hostname, Type p_type) const {
+void IP_Unix::_resolve_hostname(List<IP_Address> &r_addresses, const String &p_hostname, Type p_type) const {
 
 	struct addrinfo hints;
 	struct addrinfo *result;
@@ -105,12 +105,12 @@ void IP_Unix::_resolve_hostname(List<IP_Address> & r_addresses, const String &p_
 	int s = getaddrinfo(p_hostname.utf8().get_data(), NULL, &hints, &result);
 	if (s != 0) {
 		ERR_PRINT("getaddrinfo failed!");
-		return ;
+		return;
 	};
 
 	if (result == NULL || result->ai_addr == NULL) {
 		ERR_PRINT("Invalid response from getaddrinfo");
-		return ;
+		return;
 	};
 
 	struct addrinfo *next = result;
diff --git a/drivers/unix/ip_unix.h b/drivers/unix/ip_unix.h
index e8ead9c0a..005638aa3 100644
--- a/drivers/unix/ip_unix.h
+++ b/drivers/unix/ip_unix.h
@@ -40,7 +40,7 @@ class IP_Unix : public IP {
 	static IP *_create_unix();
 
 public:
-	virtual void _resolve_hostname(List<IP_Address> & r_addresses, const String &p_hostname, Type p_type = TYPE_ANY) const;
+	virtual void _resolve_hostname(List<IP_Address> &r_addresses, const String &p_hostname, Type p_type = TYPE_ANY) const;
 	virtual void get_local_addresses(List<IP_Address> *r_addresses) const;
 
 	static void make_default();
```

Please apply all the patches in this succession, if there is anything you disagree with, don't hesitate to argue about it :).

Overall, well done! :1st_place_medal:

**Comment by akien-mga** [comment]
```
+		if (!r_addresses.find(ip) != NULL)
+			r_addresses.push_back(_sockaddr2ip(next->ai_addr));
```

That seems a bit hard to parse for me, double negative is always tricky. Wouldn't `if (!r_addresses.find(ip))` be enough? As I parse it the current code would evaluate (when there is no match) to `if(!(bool)NULL != (bool)NULL)` i.e. `if(true != false)`, which is true, but well... :D

**Comment by Faless** [comment]
> That seems a bit hard to parse for me, double negative is always tricky. Wouldn't if (!r_addresses.find(ip)) be enough?

Nice spot, my bad, it should be `r_addresses.find(ip) == NULL` for clarity I guess. I don't much like the C style `if (ptr)` to check for NULLness. Definitely not what I wrote though.

**Comment by gotnospirit** [comment]
I did apply the patches and the hook is now installed as recommended!
Thank you, have a great christmas :)

**Review by Faless — APPROVED** [review]
Looks ready to merge for me.
Great job!

**Comment by akien-mga** [comment]
Thanks, squashed and merged manually as 010a3433d. Would it be relevant to port to the master branch too?
